### ISSUE TITLE
Rule idea: no-multi-assign

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#6424]
Proposing a new rule which forbids multiple assignments in a single expression.
#### Code that should trigger a warning:

``` js
a = b = c;
```
#### Code that should not trigger a warning:

``` js
b = c;
a = c;
```

They're hard to read and it's not immediately clear what is assigned to what. In above example, one could think that the order of evaluation is left-to-right, while it actually is right-to-left.

**Comment by ilyavolodin** [comment]
Hmm... It's a valid rule, but I'm not 100% sure it's important enough to be included in the core. We are trying to limit new rules for core to "must-haves", I'm not sure it meets the bar. @eslint/eslint-team thoughts?

**Comment by mikesherov** [comment]
I really like this as a core rule. :+1:

**Comment by michaelficarra** [comment]
@silverwind Evaluation order is very much left-to-right in assignment.

```js
z = x = {};
y = {};
x.a = x = y;
z.a === z; // true
```

**Comment by silverwind** [comment]
@michaelficarra maybe I was unclear, but it definitely assigns the right side first:

```js
o = {};
Object.defineProperty(o, "a", {set: function() { console.log("set a"); } });
Object.defineProperty(o, "b", {set: function() { console.log("set b"); } });
o.a = o.b = 1
// prints:
// set b
// set a
```

**Comment by mysticatea** [comment]
Maybe `x.a = x = y;` evaluates as the following flow.

```js
// x.a = x = y;

// save references of destinations.
d1 = &x.a
d2 = &x

// assign the value to the references by right-to-left.
*d2 = y;
*d1 = y;
```

I guess that @michaelficarra's concern is that `a = b = c` and `b = c; a = c;` are not always same.

**Comment by mysticatea** [comment]
```js
let x = {}
let obj = {
    get x() {
        console.log("get obj.x")
        return this._x
    },
    set x(value) {
        console.log("set obj.x", value)
        this._x = value
    },

    get y() {
        console.log("get obj.y")
        return this._y
    },
    set y(value) {
        console.log("set obj.y", value)
        this._y = value
    },

    _x: x,
    _y: x
}

obj.x.x = obj.y.y = obj.x = obj.y = 1

console.log(x)
```

```
get obj.x
get obj.y
set obj.y 1
set obj.x 1
{ y: 1, x: 1 }
```

**Comment by SpadeAceman** [comment]
Wow @mysticatea – that example hurts my brain!

I've been using my own custom-written version of this rule since I saw it suggested here. It's helped me identify and clean up some confusing code I'd written a while back, so to me it seems like a good core rule to have. (Using multiple assignments can seem like a good implementation of DRY, but IMO what it gains in being concise it loses in being less than clear down the road.)

**Comment by platinumazure** [comment]
This seems simple enough to implement and could be useful (as a stylistic rule).

One thing that would probably help rustle up some :+1:s from the team is if you can find a prevalent style guide or a JSCS rule which disallows multiple assignments.

**Comment by nzakas** [comment]
Unless someone steps forward to champion this rule, it's time to close this issue.

**Comment by silverwind** [comment]
There's another unmentioned risk of multiple assignment, which is the creation of unintended globals, like `b` here:

````js
let a = b = 2;
````

But I've just tried it and `no-undef` correctly identifies the issue, so that somewhat devalues this possible rule in my eyes. 

@nzakas Let's keep this issue open a few more days, and close it off as rejected if no strong case can be presented.

**Comment by mysticatea** [comment]
Actually, I don't oppose this rule. Though I like a rule which has `always`/`never` options since I think this is a stylistic rule.

If someone champions this rule, I'll support.

**Comment by SpadeAceman** [comment]
An example of an issue only this rule could catch:

```
a += b += 2 // the second += is most likely a mistake

a += b + 2 // this is likely what was intended
```

As for precedent, aside from [Python](http://effbot.org/pyfaq/why-can-t-i-use-an-assignment-in-an-expression.htm), a number of recently-created languages also treat assignments as statements, or expressions which return no value:

* Go ([section 12, Semantics: "assignment is not an expression"](https://talks.golang.org/2012/splash.article))
* Rust (["assignments always evaluate to ()" [an empty tuple]](http://stackoverflow.com/questions/38543288/rust-empty-value-for-expression))
* Swift (["the assignment operator in Swift does not itself return a value"](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/BasicOperators.html), ["Swift already deviates from C in that the =, += and other assignment-like operations return Void (for a number of reasons)."](https://github.com/apple/swift-evolution/blob/master/proposals/0004-remove-pre-post-inc-decrement.md))

While I wasn't able to find more info on *why* these these language design decisions were made, it seems significant to me that they all reject the concept of treating assignments as expressions.

**Comment by michaelficarra** [comment]
@mysticatea What would an `always` option for this rule even mean?

@SpadeAceman Thanks for the perspective. Redefining this rule as "allow assignments only in statement position" makes it much more appealing. 👍

**Comment by platinumazure** [comment]
@michaelficarra Does "statement position" also include loop initialization/afterthought and conditional tests, in your description? We would want to be very careful not to ruin for loops, for example.

**Comment by mysticatea** [comment]
@michaelficarra 

### never

```js
// ✔ GOOD
a = 1
b = 1

// ✘ BAD
a = b = 1
```

### always

```js
// ✔ GOOD
a = b = 1

// ✘ BAD
a = 1
b = 1
```

----

Ah, I had not considered about `+=`.
The rule to prevent `a += b = c`  sounds good to me.

**Comment by michaelficarra** [comment]
@platinumazure Anywhere the value is thrown away. So even `void (a = b)` is technically fine by me, and `for(a = b;;)` should be allowed, though I don't see why anyone would write that.

@mysticatea Ah, so `always` means adjacent assignments to the same primitive value or reference? I don't think that's needed.

**Comment by mysticatea** [comment]
@michaelficarra OK. 

So, I guess this is `no-multi-assign` which forbids `(node.type === "AssignmentExpression" && node.parent.type === "AssignmentExpression")`.

**Comment by platinumazure** [comment]
@mysticatea What about AssignmentExpression with parent of VariableDeclarator? That would look like a "multiple assignment" too, IMO.

**Comment by ljharb** [comment]
VariableDeclarators are asked about in #6932

**Comment by mysticatea** [comment]
@platinumazure Sounds good. `(node.type === "AssignmentExpression" && /^(?:Assignment(?:Expression|Pattern)|VariableDeclarator)$/.test(node.parent.type))`

**Comment by alberto** [comment]
@eslint/eslint-team last call for a champion or else we should close it

**Comment by platinumazure** [comment]
I'll champion.

**Comment by nzakas** [comment]
@platinumazure you need to drum up two more 👍s.

**Comment by platinumazure** [comment]
Maybe just one? I'm counting [@mysticatea](https://github.com/eslint/eslint/issues/6424#issuecomment-237520312) and [@michaelficarra](https://github.com/eslint/eslint/issues/6424#issuecomment-238294146).

@mysticatea/@michaelficarra Are you endorsing this?

**Comment by nzakas** [comment]
I can't count (I think we may need to standardize where we 👍 to make this easier). It also looks like there was a 👍 from @mikesherov and @ianvs, so marking as accepted.

**Comment by kaicataldo** [comment]
@platinumazure Are you still championing this?

**Comment by platinumazure** [comment]
Well, I want to, but with the caveat that I can't actively drive anything
here for a few weeks. I had forgotten this, probably because I forgot to
assign this to me.

We've gotten a recent request for this as well, so can we keep it open?

**Comment by kaicataldo** [comment]
I'm up for keeping it open! Just wanted to make sure it still had a champion :) Thanks!

**Comment by stewx** [comment]
I will be submitting a PR for this shortly.

**Comment by stewx** [comment]
PR here: https://github.com/eslint/eslint/pull/7904
